### PR TITLE
Speed line

### PR DESCRIPTION
**Pull request by markmellors**
speed challenge now falls back onto line following if the marker isn't seen.
tested, working.

### CODE REVIEW

File: img_base_class.py [hunk 1]
```diff
@@ -57,3 +57,64 @@ def trigger_stream(self):
...
+        # Go through each contour
+        found_area = 1
+        found_x = -1
+        found_y = -1
+        biggest_contour = None
+        for contour in contours:
+            area = cv2.contourArea(contour)
+            if found_area < area:
+                found_area = area
+                M = cv2.moments(contour)
```
Comment (robberwick) [inline img_base_class.py]: super picky minor comment: lower case variable names unless you're defining a constant or declaring a class.

File: img_base_class.py [hunk 2]
```diff
@@ -57,3 +57,64 @@ def trigger_stream(self):
...
+        return found_x, found_y, found_area, biggest_contour
+
+def wrapping_inRange(image, lower_limit, upper_limit):
+    '''function to behave like opencv imrange, but allow hue to wrap around
+    if hue in lower limit is higher than hue in upper limit, then it will use the wrapped range''' 
+    h_lower, s_lower, v_lower = lower_limit
+    h_upper, s_upper, v_upper = upper_limit
+    if h_lower > h_upper:
+        hsv_lower = (0, s_lower, v_lower)
+        hsv_upper = (h_upper, s_upper, v_upper)
```
Comment (robberwick) [inline img_base_class.py:100]: aren't you just equating `hsv_upper` to `upper_limit` here? no need to unpack/repack the tuple.

File: img_base_class.py [hunk 3]
```diff
@@ -57,3 +57,64 @@ def trigger_stream(self):
...
+        imrange1 = cv2.inRange(
+            image,
+            numpy.array(hsv_lower),
+            numpy.array(hsv_upper)
+        )
+        hsv_lower = (h_lower, s_lower, v_lower)
```
Comment (robberwick) [inline img_base_class.py:106]: another unecessary unpack/repack? `hsv_lower = lower_limit`?

File: straightline.py [hunk 4]
```diff
@@ -105,26 +119,81 @@ def process_image(self, image, screen):
...
+        screen.blit(frame, (self.LINE_CROP_TOP,0))
         pygame.display.update()
-        found_identifier = "F" if self.found else "NF"
-        img_name = "%d%simg.jpg" % (self.i, found_identifier)
+        m_found_identifier = "mF" if self.m_found else "NmF"
+        l_found_identifier = "lF" if self.l_found else "NlF"
+        img_name = "%d%s%s%dimg.jpg" % (self.i, m_found_identifier, l_found_identifier, self.ribbon_pos)
         # filesave for debugging: 
-        # cv2.imwrite(img_name, gray)
+#        if self.driving:
```
Comment (robberwick) [inline straightline.py:145]: minor: remove unused code?

File: straightline.py [hunk 5]
```diff
@@ -105,26 +119,81 @@ def process_image(self, image, screen):
...
+            image_centre_x = (self.LINE_CROP_RIGHT - self.LINE_CROP_LEFT)/2
+            t_error  = float(image_centre_x - x) / image_centre_x
+            if self.aiming:
+                turn = self.AIM_P * t_error
+            else:
+                turn = self.LINE_TURN_P * t_error
+            turn = min(max(turn,-self.MAX_TURN_SPEED), self.MAX_TURN_SPEED)
+            if self.last_t_error is not None:
+                #if there was a real error last time then do some damping
+                if self.aiming:
```
Comment (robberwick) [inline straightline.py]: the turn factor is the only thing which differs here, so reduce repetition and opportunities for future bugs:
```
turn_factor = self.LINE_TURN_D if self.aiming else self.AIM_D
turn = turn - (turn_factor * (self.last_t_error - t_error))
```
note that I'm also removing the compound assignment operator in favour of the more explicit form. Personally I'm not a fan of them, as they're overly terse, so can make scanning the code slower, and also they can trip you up when dealing with lists (i know you're not here, but it's just a bit of a bad habit to get into)

File: straightline.py [hunk 6]
```diff
@@ -174,14 +244,12 @@ def run(self):
...
         )
-        logger.info('Wait ...')
-        time.sleep(2)
         logger.info('Setting up image capture thread')
         self.image_capture_thread = ImageCapture(
             camera=self.camera,
             processor=self.processor
         )
-        pygame.mouse.set_visible(False)
+        pygame.mouse.set_visible(True)
```
Comment (robberwick) [inline straightline.py:248]: is this still true? do you still need the cursor visible?

**Review by robberwick — COMMENTED** [review]
Only minor stuff really, nothing drastic.

File: img_base_class.py (hunk 2 above)
Comment (markmellors) [reply img_base_class.py:100]: The unpack is used to make the decision and to provide the elements to repack the other tuple. Should I just be referring to elements by index/number? I could remove the second line `hsv_upper =....` and just use `upper_limit` in the imrange, but I thought this way made it clearer what was going on.

File: straightline.py (hunk 4 above)
Comment (markmellors) [reply straightline.py:145]: I should make a command line parameter to enable image saving, as this comes up in every challenge.

File: straightline.py (hunk 6 above)
Comment (markmellors) [reply straightline.py:248]: cursor is now used to indicate perceived line or marker position

File: img_base_class.py (hunk 1 above)
Comment (markmellors) [reply img_base_class.py]: this was from open CV demo code, and M makes more sense for moments for me than m (usually mass). I've changed it keep to code.

**Pull request merged**